api_tests/harness.js

```diff
@@ -16,202 +20,267 @@ const project_root = execSync("git rev-parse --show-toplevel", {
+        this.running_ledgers = {};
+        this.ledgers_config = ledgers_config;
+        this.block_timers = {};
+    }
+
+    async ensure_ledgers_running(ledgers) {
+        let wait_times = [0];
+        let running_ledgers = this.running_ledgers;
+
+        let to_be_started = ledgers.filter(name => !running_ledgers[name]);
+
+        if (to_be_started.length > 0) {
+            for (let ledger of to_be_started) {
+                let ledger_config = ledgers_config[ledger];
+
+                await execSync(
+                    "docker-compose up -d " + ledger_config.docker,
+                    docker_compose_options
+                );
+
+                let log_process = await spawn(
+                    "docker-compose",
+                    ["logs", "--tail=all", "-f"],
+                    {
+                        cwd: docker_cwd,
+                        encoding: "utf-8",
+                        stdio: [
+                            "ignore",
+                            fs.openSync(log_dir + "/docker-compose.log", "w"),
+                            fs.openSync(
+                                log_dir + "/docker-compose-err.log",
+                                "w"
+                            ),
+                        ],
+                    }
+                );
+
+                this.running_ledgers[ledger] = { log_process };
+                wait_times.push(parseInt(ledger_config.up_time));
+            }
 
-async function startDockerContainers(ledgers, ledgers_config) {
-    if (ledgers.length === 0) {
-        throw new Error("No ledgers to start");
+            if (to_be_started.includes("bitcoin")) {
+                this.block_timers.bitcoin = setInterval(async () => {
+                    await bitcoin.btc_generate();
+                }, 3000);
+            }
+
+            let wait_time = Math.max(...wait_times);
+            console.log(
+                "Waiting " +
+                    wait_time +
+                    " for " +
+                    to_be_started.join(", ") +
+                    " to start"
+            );
+            await util.sleep(wait_time);
+        }
     }
 
-    let names = [];
-    ledgers.forEach(function(ledger) {
-        names.push(ledgers_config[ledger].docker);
-    });
-    await execSync(
-        "docker-compose up -d " + names.join(" "),
-        docker_compose_options
-    );
-
-    return await spawn("docker-compose", ["logs", "--tail=all", "-f"], {
-        cwd: docker_cwd,
-        encoding: "utf-8",
-        stdio: [
-            "ignore",
-            fs.openSync(log_dir + "/docker-compose.log", "w"),
-            fs.openSync(log_dir + "/docker-compose-err.log", "w"),
-        ],
-    });
+    stop_ledgers() {
+        let names = Object.keys(this.running_ledgers);
+        if (names.length > 0) {
+            console.log("Stopping ledgers: " + names.join(", "));
+            for (let [name, block_timer] of Object.entries(this.block_timers)) {
```

Comment:
:wrench: This is a bit more concise:

```js
Object.values(this.block_timers).forEach(clearInterval)
```

Reply:
❌ please don't use that as it is not TypeScript friendly :)

Reply:
I did the thing @thomaseizinger suggested will that work?